graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py

```diff
@@ -0,0 +1,81 @@
+from collections import OrderedDict
+import warnings
+
+from graphql.type import GraphQLBoolean, GraphQLFloat, GraphQLInt, GraphQLString
+import six
+import sqlalchemy.sql.sqltypes as sqltypes
+
+from ...schema import GraphQLDate, GraphQLDateTime, GraphQLDecimal
+from ..schema_graph import PropertyDescriptor, SchemaGraph, VertexType
+
+
+# TODO(pmantica1): Add scalar mapping for the following classes: Interval, and Time.
+# We do not currently plan to add a mapping for JSON and Binary objects.
+SQL_CLASS_TO_GRAPHQL_TYPE = OrderedDict([
+    (sqltypes.String, GraphQLString),
+    (sqltypes.Integer, GraphQLInt),
+    (sqltypes.Float, GraphQLFloat),
+    (sqltypes.Numeric, GraphQLDecimal),
+    (sqltypes.DateTime, GraphQLDateTime),
+    (sqltypes.Date, GraphQLDate),
+    (sqltypes.Boolean, GraphQLBoolean),
+])
+
+
+# TODO(pmantica1): Map foreign keys to edges.
+# TODO(pmantica1): Represent table inheritance in SchemaGraph.
+# TODO(pmantica1): Add option to map tables to EdgeTypes instead of VertexTypes.
+# TODO(pmantica1): Map arrays to GraphQLLists.
+def get_schema_graph_from_sql_alchemy_metadata(sqlalchemy_metadata):
+    """Return the matching SchemaGraph for the SQLAlchemy Metadata object"""
+    _validate_sql_to_graphql_is_toposorted_by_class_inheritance()
+    elements = dict()
+    for table_name, table in six.iteritems(sqlalchemy_metadata.tables):
+        elements[table_name] = _get_vertex_type_from_sqlalchemy_table(table)
+    inheritance_sets = {element_name: {element_name} for element_name in elements}
+    return SchemaGraph(elements, inheritance_sets)
+
+
+def _validate_sql_to_graphql_is_toposorted_by_class_inheritance():
+    """Validate that SQL_SCALAR_CLASS_TO_GRAPHQL_TYPE dict is toposorted by class inheritance."""
+    sql_classes = list(SQL_CLASS_TO_GRAPHQL_TYPE.keys())
+    for i, class_ in enumerate(sql_classes):
+        for other_class in sql_classes[i + 1:]:
+            if issubclass(other_class, class_):
+                raise AssertionError('SQL_CLASS_TO_GRAPHQL_TYPE is not toposorted by class '
+                                     'inheritance. Class {} is a subclass of {} but appears '
+                                     'later in the OrderedDict.'.format(other_class, class_))
+
+
+def _try_get_graphql_scalar_type(column_name, column_type):
+    """Return the most precise GraphQLScalarType for the SQL datatype class.
+
+    For instance, if class of the column_type is the Numeric class we return GraphQLDecimal.
+    If the class of the column_type is Float, a subclass of Numeric, we return GraphQLFloat.
+    """
+    maybe_graphql_type = None
+    for sql_class, graphql_type in SQL_CLASS_TO_GRAPHQL_TYPE.items():
+        if isinstance(column_type, sql_class):
+            maybe_graphql_type = graphql_type
+            break
+    if not maybe_graphql_type:
```

Comment:
can you make a list of the unsupported ones, and check if it's there? Then if it's not either supported or unsupported, you can raise an `AssertionError`. This way we will know exactly what's unsupported.

Reply:
I think we should warn over raising an AssertionError like we currently do. I have that list in the tests.

Reply:
The list is in the test. I don't think we should raise an AssertionError because we won't be able to generate a schema then unless we support all the types. I think it's best to just ignore some columns.

Reply:
The `AssertionError` would be raised if we get a type that's neither supported nor unsupported. This can happen if we fail to account for all types that exist. For example, it would be really sad if SQL had 2 different types of ints, and we only supported one, without knowing that we just dropped all ints of the second type.

Reply:
@bojanserafimov would it help if our tests are run in such a way that warnings are treated as exceptions, so that they are guaranteed to fail if this warning is hit in the test environment?

What I am a bit worried about is this scenario:
- SQLAlchemy updates the SQL types it exposes, adding a new type that is not in our list.
- The compiler breaks with that version of SQLAlchemy, and we have to release a new patch on an emergency basis that updates that list to make it explicitly unsupported.

Reply:
This makes sense to me. @bojanserafimov what do you think?

Reply:
I see, that would be bad. Agree.

A slight problem with this is that tests will never hit that warning because they only test with a hard-coded list of types. If we can assert that the types are what we think we are, that would be great. But I don't see a way to do that.

Reply:
Ok, I added a warning test.